Packet++/src/VrrpLayer.cpp

```diff
@@ -0,0 +1,533 @@
+			return;
+		}
+
+		if (ipAddress.isIPv4()) {
+			memcpy(data, ipAddress.getIPv4().toBytes(), ipAddressLen);
+		} else if (ipAddress.isIPv6()) {
+			memcpy(data, ipAddress.getIPv6().toBytes(), ipAddressLen);
+		} else {
+		}
+	}
+
+	bool VrrpLayer::getIPAddressFromData(uint8_t *data, IPAddress &ipAddress) const {
+		if (getAddressType() == IPAddress::IPv4AddressType) {
+			IPv4Address ipv4Address(*((uint32_t *) data));
+			ipAddress = ipv4Address;
+		} else if (getAddressType() == IPAddress::IPv6AddressType) {
+			IPv6Address ipv6Address(data);
+			ipAddress = ipv6Address;
+		} else {
+			return false;
+		}
+
+		return true;
+	}
+
+	bool VrrpLayer::isIPAddressValid(IPAddress &ipAddress) const {
+		if (ipAddress.isIPv6() && (getProtocol() != VRRPv3)) {
+			PCPP_LOG_ERROR("Only VRRPv3 support IPv6 virtual address.");
+			return false;
+		}
+		if (ipAddress.getType() != getAddressType()) {
+			PCPP_LOG_ERROR("IP address version is not equal to layer's.");
+			return false;
+		}
+		if (!ipAddress.isValid()) {
+			PCPP_LOG_ERROR("IP address is invalid.");
+			return false;
+		}
+
+		return true;
+	}
+
+	IPAddress::AddressType VrrpLayer::getAddressType() const {
+		return m_addressType;
+	}
+
+	void VrrpLayer::setAddressType(IPAddress::AddressType addressType) {
+		m_addressType = addressType;
+	}
+
+	/*************
+	 * Vrrpv2Layer
+	 *************/
+
+	std::string VrrpV2Layer::getAuthTypeDesc() const {
+		std::string toStr;
+
+		toStr += "\n\tAuth Type: " + getAuthTypeDescByType(getAuthType());
+		toStr += " (";
+		toStr += std::to_string(getAuthType());
+		toStr += ")";
+
+		return toStr;
+	}
+
+	uint8_t VrrpV2Layer::getAdvInt() const {
+		auto *packet = (vrrp_packet *) getData();
+		uint16_t authAdvInt = packet->authTypeAdvInt;
+		auto *authAdvIntPtr = (vrrpv2_auth_adv *) (&authAdvInt);
+		return authAdvIntPtr->advInt;
+	}
+
+	uint8_t VrrpV2Layer::getAuthType() const {
+		auto *packet = (vrrp_packet *) getData();
+		uint16_t authAdvInt = packet->authTypeAdvInt;
+		auto *authAdvIntPtr = (vrrpv2_auth_adv *) (&authAdvInt);
+		return authAdvIntPtr->authType;
+	}
+
+	uint16_t VrrpV2Layer::fillChecksum() {
+		auto *vrrpPacket = (vrrp_packet *) getData();
+		ScalarBuffer<uint16_t> buffer = {};
+		buffer.buffer = (uint16_t *) getData();
+		buffer.len = getHeaderLen();
+
+		vrrpPacket->checksum = 0;
+		uint16_t checksum = computeChecksum(&buffer, 1);
+		vrrpPacket->checksum = htobe16(checksum);
+
+		return checksum;
+	}
+
+	uint16_t VrrpV2Layer::calculateChecksum() const {
+		if ((getData() == nullptr) || (getDataLen() == 0)) {
+			return 0;
+		}
+
+		auto *vrrpPacketCopy = (vrrp_packet *) new uint8_t[getDataLen()];
+		memcpy(vrrpPacketCopy, getData(), getDataLen());
+
+		ScalarBuffer<uint16_t> buffer = {};
+		buffer.buffer = (uint16_t *) vrrpPacketCopy;
+		buffer.len = getHeaderLen();
+
+		vrrpPacketCopy->checksum = 0;
+		uint16_t checksum = computeChecksum(&buffer, 1);
+		delete[] vrrpPacketCopy;
```

Comment:
coping the data is quite expensive. Instead we can save the existing checksum value in a variable, set the checksum to 0, calculate the checksum and then set the value back. This is what we do in `TcpLayer`:
https://github.com/seladb/PcapPlusPlus/blob/3dca454eaf8f8440c3e302df262dc5da637b6c5f/Packet%2B%2B/src/TcpLayer.cpp#L246

Reply:
OK. I will fix it.